### PR TITLE
Updating the Security Baseline checks for SSH and adding automatic remediation 

### PR DESCRIPTION
**Pull request by MariusNi**
## Description

This is the first part of the work for SSH Server policy. The SSH Baseline Security checks are updated as stated in the design and remediation added to the Security Baseline before the SSH part will be separated. 

## Checklist

- [X] I have read the [contribution guidelines](https://github.com/Azure/azure-osconfig/blob/main/CONTRIBUTING.md).
- [X] I added unit-tests to validate my changes. All unit tests are passing.
- [X] I have merged the latest `main` branch prior to this PR submission.
- [X] I submitted this PR against the `main` branch.

### CODE REVIEW

File: src/common/commonutils/DeviceInfoUtils.c [hunk 1]
```diff
@@ -48,6 +48,24 @@ void RemovePrefixUpTo(char* target, char marker)
...
 }
 
+void RemovePrefixUpToString(char* target, const char* marker)
+{
+    if ((NULL == target) || (NULL == marker))
+    {
+        return;
+    }
+
+    int targetLength = (int)strlen(target);
```
Comment (simonjaeger) [inline src/common/commonutils/DeviceInfoUtils.c:58]: nit: Seems like this assignment can be removed (just be set to zero). It is reassigned again below (line 63) before first usage.

File: src/common/commonutils/CommonUtils.h [hunk 2]
```diff
@@ -100,11 +100,11 @@ int FindTextInCommandOutput(const char* command, const char* text, char** reason
 int CheckLockoutForFailedPasswordAttempts(const char* fileName, void* log);
 int CheckOnlyApprovedMacAlgorithmsAreUsed(const char** macs, unsigned int numberOfMacs, char** reason, void* log);
 int CheckAppropriateCiphersForSsh(const char** ciphers, unsigned int numberOfCiphers, char** reason, void* log);
-int CheckLimitedUserAcccessForSsh(const char** values, unsigned int numberOfValues, char** reason, void* log);
-int CheckSshOptionIsSetToString(const char* option, const char* expectedValue, char** reason, void* log);
-int CheckSshOptionIsSetToInteger(const char* option, int expectedValue, int* actualValue, char** reason, void* log);
-int CheckSshIdleTimeoutInterval(char** reason, void* log);
+int CheckSshOptionIsSet(const char* option, const char* expectedValue, char** actualValue, char** reason, void* log);
```
Comment (AhmedBM) [inline src/common/commonutils/CommonUtils.h:103]: Should't it be `CheckSSHOptionIsSet`? Looks better that way and also properly pascal cased.

File: src/common/commonutils/SshUtils.c [hunk 3]
```diff
@@ -220,171 +235,180 @@ int CheckAppropriateCiphersForSsh(const char** ciphers, unsigned int numberOfCip
...
+    FREE_MEMORY(actualValueString);
+    FREE_MEMORY(expectedValueString);
+
+    return status;
+}
+
+int CheckSshClientAliveInterval(char** reason, void* log)
+{
+    int actualValue = 0;
+    int status = CheckSshOptionIsSetToInteger("clientaliveinterval", NULL, &actualValue, reason, log);
```
Comment (AhmedBM) [inline src/common/commonutils/SshUtils.c:305]: Can we turn `clientaliveinterval` into a constant?

File: src/common/commonutils/SshUtils.c [hunk 4]
```diff
@@ -220,171 +235,180 @@ int CheckAppropriateCiphersForSsh(const char** ciphers, unsigned int numberOfCip
...
+    return status;
+}
 
-        FREE_MEMORY(value);
-    }
-    else
+int CheckSshLoginGraceTime(char** reason, void* log)
+{
+    int actualValue = 0;
+    int status = CheckSshOptionIsSetToInteger("logingracetime", NULL, &actualValue, reason, log);
```
Comment (AhmedBM) [inline src/common/commonutils/SshUtils.c:323]: Should turn `logingracetime` into a constant

File: src/common/commonutils/SshUtils.c [hunk 5]
```diff
@@ -220,171 +235,180 @@ int CheckAppropriateCiphersForSsh(const char** ciphers, unsigned int numberOfCip
...
     return status;
 }
 
-int CheckSshOptionIsSetToInteger(const char* option, int expectedValue, int* actualValue, char** reason, void* log)
+int SetSshOption(const char* option, const char* value, void* log)
 {
-    char* value = NULL;
-    int integerValue = 0;
+    // Replaces any instances of 'option foo' with 'option value' and adds 'option value' when 'option' is missing
+    const char* commandTemplate = "sed '/^%s /{h;s/ .*/ %s/};${x;/^$/{s//%s %s/;H};x}' %s";
```
Comment (AhmedBM) [inline src/common/commonutils/SshUtils.c:341]: is `sed` available in all our supported distros?

**Review by AhmedBM — DISMISSED** [review]
Just a few minor comments but looks great overall.

File: src/common/commonutils/DeviceInfoUtils.c (hunk 1 above)
Comment (MariusNi) [reply src/common/commonutils/DeviceInfoUtils.c:58]: Well spotted.

File: src/common/commonutils/CommonUtils.h (hunk 2 above)
Comment (MariusNi) [reply src/common/commonutils/CommonUtils.h:103]: 3 letter acronyms are cased and is also the common pattern in this project.

File: src/common/commonutils/SshUtils.c (hunk 3 above)
Comment (MariusNi) [reply src/common/commonutils/SshUtils.c:304]: It's a constant. In general, where the constant is used in one place, it's coded in there. If it used in multiple places, it's defined like a real constant.

File: src/common/commonutils/SshUtils.c (hunk 4 above)
Comment (MariusNi) [reply src/common/commonutils/SshUtils.c:322]: Same as commented elsewhere.

File: src/common/commonutils/SshUtils.c (hunk 5 above)
Comment (MariusNi) [reply src/common/commonutils/SshUtils.c:340]: Yes. It's using on purpose not complications such as -i.

**Comment by MariusNi** [comment]
Thanks Ahmed!

---
In reply to: [1725720106](https://github.com/Azure/azure-osconfig/pull/572#issuecomment-1725720106) [](http://example.com/codeflow?ancestors=1725720106)

**Review by AhmedBM — APPROVED** [review]
👍

**Review by robertschaedler3 — APPROVED** [review]

**Pull request merged**